etl/src/main/java/org/pih/hivmigration/etl/sql/ProgramMigrator.groovy

```diff
@@ -0,0 +1,252 @@
+                starting_health_center,
+                health_center,
+                health_center_transfer_date,
+                treatment_status,
+                treatment_status_date,
+                art_start_date,
+                regimen_outcome,
+                regimen_outcome_date,
+                min_visit_date,
+                max_visit_date)
+            values (?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?)
+        ''', '''
+            select 	
+                d.patient_id as source_patient_id,
+                d.starting_health_center,
+                d.health_center,
+                d.health_center_transfer_date,
+                d.treatment_status,
+                d.treatment_status_date,
+                r.start_date,
+                o.regimen_outcome,
+                o.regimen_outcome_date,
+                e.min_visit_date,
+                e.max_visit_date
+            from
+                hiv_demographics_real d, hiv_art_regimen_start_date r, hiv_reg_outcome_helper_dates o,
+                (
+                    select 		patient_id, min(encounter_date) as min_visit_date, max(encounter_date) as max_visit_date 
+                    from 		hiv_encounters 
+                    where 		type in ('intake','followup','patient_contact','accompagnateur') 
+                    and			encounter_date is not null
+                    group by 	patient_id
+                ) e
+            where
+                d.TREATMENT_STATUS != 'test' 
+            and 
+                d.patient_id = e.patient_id
+            and
+                d.patient_id = r.patient_id(+)
+            and
+                d.patient_id = o.patient_id(+)
+        ''')
+
+        executeMysql("Add enrollment_date and outcome", '''
+            UPDATE hivmigration_programs_raw
+            SET
+                # Enrollment Date in program is earlier of minimum encounter and art start date
+                enrollment_date = LEAST (IFNULL(min_visit_date, art_start_date), IFNULL(art_start_date, min_visit_date)),                
+                outcome = CASE treatment_status
+                    WHEN 'died' THEN 'DIED'
+                    WHEN 'lost' THEN 'LOST' 
+                    WHEN 'abandoned' THEN 'LOST'
+                    WHEN 'transferred_out' THEN 'TRANSFERRED_OUT'
+                    WHEN 'treatment_refused' THEN 'TREATMENT_STOPPED'  # TODO: this might be its own outcome
+                    WHEN 'treatment_stopped' THEN 'TREATMENT_STOPPED'
+                    WHEN 'treatment_stopped_side_effects' THEN 'TREATMENT_STOPPED'
+                    WHEN 'treatment_stopped_other' THEN 'TREATMENT_STOPPED'
+                    END
+            ;
+        ''')
+
+
+        executeMysql("Add outcome date if outcome present", '''
+            UPDATE hivmigration_programs_raw
+            SET
+                # Outcome date is the most recent of treatment_status_date, regimen_outcome_date, and enrollment date
+                outcome_date = IFNULL(
```

Comment:
I can't follow this nearly as easily as I could follow a series of update statements.  Something like:

```
update hivmigration_programs_raw set outcome_date = treatment_status_date;

update hivmigration_programs_raw set outcome date = regimen_outcome_date where outcome_date is null or regimen_outcome_date > outcome_date;

update hivmigration_programs_raw set outcome_date = enrollment_date where outcome_date is null or enrollment_date > outcome_date;  // I'm not sure why we would want this, but ok

update hivmigration_programs_raw set outcome_date = date_add(max_visit_date, INTERVAL 6 MONTH) where outcome is null;

update hivmigration_programs_raw set outcome_date = today() where outcome_date > today();
...

```

Reply:
Again, shouldn't be a blocker on merging this in if it's migrating existing logic, but worth documenting and reviewing.

Reply:
And specifically, setting outcome date to entrollment date does seem weird as @mseaton  mentioned above.  Also, are we saying there's always an outcome date?  ie, all programs should be closed? that doesn't seem to make sense?